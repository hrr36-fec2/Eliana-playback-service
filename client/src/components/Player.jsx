import React from 'react';
import { createPlaylist, playlist } from '../playlist.js';
import MetaData from './MetaData.jsx';

class Player extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      id: null,
      trackId: null,
      title: '',
      artist: '',
      albumThumbnail: '',
      favorite: 0,
      isplaying: false,
      sourceURL: '',
      previous: null,
      next: null,
      progressbar: 0,
      currentTime: '00:00',
      duration: '00:00',
      loop: false,
      shuffle: false
    };

    this.player = new Audio();
    this.playTrack = this.playTrack.bind(this);
    this.startPlayer = this.startPlayer.bind(this);
    this.playOrPause = this.playOrPause.bind(this);
    this.playNext = this.playNext.bind(this);
    this.playPrev = this.playPrev.bind(this);
    // this.favorite = this.favorite.bind(this);

  }

  //if track is requested by a separate component in scootify 
  playTrack(trackId) {
    fetch('http://localhost:3001/track/' + trackId, {
      method: 'GET'
    })
      .then(res => res.json())
      .then(
        (result) => {
          console.log(result, 'SUCCESS GETTING TRACK ' + trackId);
          this.setState({
            id: result[0].id,
            trackId: trackId,
            isplaying: true,
            sourceURL: result[0].track_file_url,
            title: result[0].track_title,
            favorite: result[0].favorite,
            artist: result[0].artist_name,
            albumThumbnail: result[0].album_image_file,
          });
        })
      .then(() => {
        this.player.src = this.state.sourceURL;
        this.player.play();
      })
      .catch(
        (error) => {
          console.log(error, 'ERROR FETCHING TRACK ' + trackID);
        }
      );
  }

  startPlayer(index) {
    this.setState({
      id: playlist[index].id,
      trackId: playlist[index].track_id,
      sourceURL: playlist[index].track_file_url,
      title: playlist[index].track_title,
      artist: playlist[index].artist_name,
      albumThumbnail: playlist[index].album_image_file,
      favorite: playlist[index].favorite
    });
    if (this.state.previous === null) {
      this.setState({ previous: index });
    }
    if (this.state.next === null) {
      this.setState({ next: index + 1 });
    }
  }

  playOrPause() {
    if (this.player.paused) {
      this.player.src = this.state.sourceURL;
      this.player.play();
      this.setState({ isplaying: true });
    } else {
      this.player.pause();
      this.setState({ isplaying: false });
    }
  }

  playNext() {
    if (this.state.shuffle) {
      this.startPlayer(Math.floor(Math.random() * playlist.length));
    } else {
      if (this.state.index === playlist.length - 1) {
        this.player.pause();
        this.setState({ isplaying: false });
      } else {
        this.setState({ next: this.state.index + 1 });
        this.startPlayer(this.state.next);
      }
    }
  }

  playPrev() {
    if (this.state.currentTime <= '00:03') {
      this.startPlayer(this.state.id - 1);
    }
    if (this.state.previous >= 0) {
      this.startPlayer(this.state.previous);
    } else {
      this.player.pause();
      this.setState({ isplaying: false });
    }
  }

  componentDidMount() {
    createPlaylist((error) => {
      if (error) {
        console.log(error, 'ERROR IN COMPONENTDIDMOUNT');
      } else {
        console.log(playlist, 'THIS IS THE PLAYLIST');
        this.startPlayer(0);
      }
    });
  }

  render() {
    return (
      <div id="player">
        <MetaData
          title={this.state.title}
          artist={this.state.artist}
          albumThumbnail={this.state.albumThumbnail}
          favorite={this.state.favorite}
        />
        <div className="middle-part">
          <div className="controls">
            <button className={this.state.shuffle ? "no-style shuffle active" : "no-style shuffle"} onClick={this.toggleShuffle}><i className="fa fa-random"></i></button>
            <button className="no-style" onClick={this.playPrev}><i className="fa fa-step-backward"></i></button>
            <button className="no-style play" onClick={this.playOrPause}>
              <i className={this.state.isplaying ? "far fa-pause-circle" : "far fa-play-circle"}></i>
            </button>
            <button className="no-style" onClick={this.playNext}><i className="fa fa-step-forward"></i></button>
            <button className={this.state.loop ? "no-style loop active" : "no-style loop"} onClick={this.toggleLoop}><i className="fa fa-sync-alt"></i></button>
          </div>
        </div>
      </div>
    );
  }
}

export default Player;