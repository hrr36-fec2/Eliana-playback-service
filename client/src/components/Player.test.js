import React from 'react';
import { shallow, mount } from 'enzyme';
import Player from './Player';

describe('Player', () => {
  it('should render correctly', () => {
    const component = shallow(<Player />);
    expect(component).toMatchSnapshot();
  });

  it('should render metaData', () => {
    const wrap = mount(<Player />);
    wrap.update();
    expect(wrap).toMatchSnapshot();
    wrap.unmount();
  });
});

