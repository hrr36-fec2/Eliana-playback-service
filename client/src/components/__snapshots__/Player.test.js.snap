// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player should render correctly 1`] = `
<div
  id="player"
>
  <MetaData
    albumThumbnail={null}
    artist={null}
    favorite={0}
    onClick={[Function]}
    title={null}
  />
  <div
    className="now-playing-bar__center"
  >
    <div
      className="controls"
    >
      <button
        className="no-style shuffle"
        onClick={[Function]}
      >
        <ion-icon
          name="shuffle"
        />
      </button>
      <button
        className="no-style"
        onClick={[Function]}
      >
        <ion-icon
          name="skip-backward"
        />
      </button>
      <button
        className="no-style play"
        onClick={[Function]}
      >
        <i
          className="far fa-play-circle"
        />
      </button>
      <button
        className="no-style"
        onClick={[Function]}
      >
        <ion-icon
          name="skip-forward"
        />
      </button>
      <button
        className="no-style loop"
        onClick={[Function]}
      >
        <ion-icon
          name="repeat"
        />
      </button>
    </div>
    <div
      className="progress"
    >
      <div
        className="progress-time"
      >
        00:00
      </div>
      <div
        className="bar"
        onClick={[Function]}
      />
      <div
        className="progress-time"
      >
        00:00
      </div>
    </div>
  </div>
  <div
    className="volume-bar"
  >
    <button
      className="no-style"
      onClick={[Function]}
    >
      <i
        className="fa fa-volume-up"
      />
    </button>
    <div
      className="fill"
      onClick={[Function]}
    >
      <div
        style={
          Object {
            "width": "100%",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`Player should render metaData 1`] = `
<Player>
  <div
    id="player"
  >
    <MetaData
      albumThumbnail={null}
      artist={null}
      favorite={0}
      onClick={[Function]}
      title={null}
    >
      <div
        className="now-playing"
      >
        <img
          alt="album cover art thumbnail"
          src={null}
        />
        <div
          className="meta"
        >
          <span
            className="title"
          />
          <span
            className="artist-name"
          />
        </div>
        <button
          className="favorite"
          onClick={[Function]}
        >
          <i
            className="far fa-heart"
          />
        </button>
      </div>
    </MetaData>
    <div
      className="now-playing-bar__center"
    >
      <div
        className="controls"
      >
        <button
          className="no-style shuffle"
          onClick={[Function]}
        >
          <ion-icon
            name="shuffle"
          />
        </button>
        <button
          className="no-style"
          onClick={[Function]}
        >
          <ion-icon
            name="skip-backward"
          />
        </button>
        <button
          className="no-style play"
          onClick={[Function]}
        >
          <i
            className="far fa-play-circle"
          />
        </button>
        <button
          className="no-style"
          onClick={[Function]}
        >
          <ion-icon
            name="skip-forward"
          />
        </button>
        <button
          className="no-style loop"
          onClick={[Function]}
        >
          <ion-icon
            name="repeat"
          />
        </button>
      </div>
      <div
        className="progress"
      >
        <div
          className="progress-time"
        >
          00:00
        </div>
        <div
          className="bar"
          onClick={[Function]}
        />
        <div
          className="progress-time"
        >
          00:00
        </div>
      </div>
    </div>
    <div
      className="volume-bar"
    >
      <button
        className="no-style"
        onClick={[Function]}
      >
        <i
          className="fa fa-volume-up"
        />
      </button>
      <div
        className="fill"
        onClick={[Function]}
      >
        <div
          style={
            Object {
              "width": "100%",
            }
          }
        />
      </div>
    </div>
  </div>
</Player>
`;
